Mathlib/Algebra/MonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,2001 @@
+    IsScalarTower R S (MonoidAlgebra k G) :=
+  Finsupp.isScalarTower G k
+#align monoid_algebra.is_scalar_tower MonoidAlgebra.isScalarTower
+
+instance smulCommClass [Monoid R] [Monoid S] [Semiring k] [DistribMulAction R k]
+    [DistribMulAction S k] [SMulCommClass R S k] : SMulCommClass R S (MonoidAlgebra k G) :=
+  Finsupp.smulCommClass G k
+#align monoid_algebra.smul_comm_tower MonoidAlgebra.smulCommClass
+
+instance isCentralScalar [Monoid R] [Semiring k] [DistribMulAction R k] [DistribMulAction Rᵐᵒᵖ k]
+    [IsCentralScalar R k] : IsCentralScalar R (MonoidAlgebra k G) :=
+  Finsupp.isCentralScalar G k
+#align monoid_algebra.is_central_scalar MonoidAlgebra.isCentralScalar
+
+/-- This is not an instance as it conflicts with `MonoidAlgebra.distribMulAction` when `G = kˣ`.
+-/
+def comapDistribMulActionSelf [Group G] [Semiring k] : DistribMulAction G (MonoidAlgebra k G) :=
+  Finsupp.comapDistribMulAction
+#align monoid_algebra.comap_distrib_mul_action_self MonoidAlgebra.comapDistribMulActionSelf
+
+end DerivedInstances
+
+section MiscTheorems
+
+variable [Semiring k]
+
+-- attribute [local reducible] MonoidAlgebra -- Porting note: `reducible` cannot be `local`.
+
+theorem mul_apply [DecidableEq G] [Mul G] (f g : MonoidAlgebra k G) (x : G) :
+    (f * g) x = f.sum fun a₁ b₁ => g.sum fun a₂ b₂ => if a₁ * a₂ = x then b₁ * b₂ else 0 := by
+  rw [mul_def]
+  simp only [Finsupp.sum_apply, single_apply]
+#align monoid_algebra.mul_apply MonoidAlgebra.mul_apply
+
+theorem mul_apply_antidiagonal [Mul G] (f g : MonoidAlgebra k G) (x : G) (s : Finset (G × G))
+    (hs : ∀ {p : G × G}, p ∈ s ↔ p.1 * p.2 = x) : (f * g) x = ∑ p in s, f p.1 * g p.2 := by
+  classical exact
+      let F : G × G → k := fun p => if p.1 * p.2 = x then f p.1 * g p.2 else 0
+      calc
+        (f * g) x = ∑ a₁ in f.support, ∑ a₂ in g.support, F (a₁, a₂) := mul_apply f g x
+        _ = ∑ p in f.support ×ᶠ g.support, F p := Finset.sum_product.symm
+        _ = ∑ p in (f.support ×ᶠ g.support).filter fun p : G × G => p.1 * p.2 = x, f p.1 * g p.2 :=
+          (Finset.sum_filter _ _).symm
+        _ = ∑ p in s.filter fun p : G × G => p.1 ∈ f.support ∧ p.2 ∈ g.support, f p.1 * g p.2 :=
+          (sum_congr
+            (by
+              ext
+              simp only [mem_filter, mem_product, hs, and_comm])
+            fun _ _ => rfl)
+        _ = ∑ p in s, f p.1 * g p.2 :=
+          sum_subset (filter_subset _ _) fun p hps hp =>
+            by
+            simp only [mem_filter, mem_support_iff, not_and, Classical.not_not] at hp⊢
+            by_cases h1 : f p.1 = 0
+            · rw [h1, zero_mul]
+            · rw [hp hps h1, mul_zero]
+
+#align monoid_algebra.mul_apply_antidiagonal MonoidAlgebra.mul_apply_antidiagonal
+
+-- Porting note: Type ascriptions didn't work, so `α` & `β` of `HMul.hMul` are specified.
+@[simp]
+theorem single_mul_single [Mul G] {a₁ a₂ : G} {b₁ b₂ : k} :
+    HMul.hMul (α := MonoidAlgebra k G) (β := MonoidAlgebra k G) (single a₁ b₁) (single a₂ b₂)
```

Comment:
Hmm, that's a pity. We should try to solve this, or at least create an issue for it.
That issue should then be mentioned in all places in this file where the problem occurs.

Reply:
Mario suggested in a PM that we should define `MonoidAlgebra.single` that has the right type.

Reply:
I've submited [the issue](https://github.com/leanprover/lean4/issues/2134).

Reply:
Could you please also define `MonoidAlgebra.single` and use that instead of `Finsupp.single`? Then you should be able to use notation as usual, without passing the types explicitly.

Reply:
Even I define `MonoidAlgebra.single` as `abbrev`, it caused many problems in `simp`s. Instead, can I introduce these local notations?
```lean
local infixl:70 " *ₘ "   =>
  @HMul.hMul _ _ _ (inferInstanceAs (HMul (MonoidAlgebra _ _) (MonoidAlgebra _ _) _))

local infixl:80 " ^ₘ "   =>
  @HPow.hPow _ _ _ (inferInstanceAs (HPow (MonoidAlgebra _ _) _ _))

local infixr:73 " •ₘ " =>
  @HSMul.hSMul _ _ _ (inferInstanceAs (HSMul (MonoidAlgebra _ _) _ _))

local syntax (name := bigprodin) "∏ₘ " Std.ExtendedBinder.extBinder "in " term "," term:67 : term

macro_rules
  | `(∏ₘ $x:ident in $s, $r) =>
    `(@Finset.prod _ _ (inferInstanceAs (CommMonoid (MonoidAlgebra _ _))) $s (fun $x ↦ $r))
  | `(∏ₘ $x:ident : $t in $s, $p) =>
    `(Finset.prod _ _ (inferInstanceAs (CommMonoid (MonoidAlgebra _ _)))
      $s (fun $x:ident : $t ↦ $p))
```

Reply:
@digama0 what do you suggest?

Reply:
I suggest defining `MonoidAlgebra.single` and copying every lemma about `finsupp.single` to `MonoidAlgebra.single`.